caddytls/certificate_transparency.go

```diff
@@ -0,0 +1,122 @@
+package caddytls
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// This code borrows heavily from https://github.com/grahamedgecombe/ct-submit
+
+type addChainRequest struct {
+	Chain []string `json:"chain"`
+}
+
+type signedCertificateTimestamp struct {
+	Version    uint8  `json:"sct_version"`
+	LogID      string `json:"id"`
+	Timestamp  int64  `json:"timestamp"`
+	Extensions string `json:"extensions"`
+	Signature  string `json:"signature"`
+}
+
+// AsRawBytes converts the SCT from the JSON form that the submission API
+// returns into the raw binary form that the TLS handshake expects.
+func (sct *signedCertificateTimestamp) AsRawBytes() ([]byte, error) {
+	b := new(bytes.Buffer)
+	// Version
+	binary.Write(b, binary.BigEndian, sct.Version)
+
+	// LogID
+	logBytes, err := base64.StdEncoding.DecodeString(sct.LogID)
+	if err != nil {
+		return nil, err
+	}
+	b.Write(logBytes)
+
+	// Timestamp
+	binary.Write(b, binary.BigEndian, sct.Timestamp)
+
+	// Extensions
+	extensionBytes, err := base64.StdEncoding.DecodeString(sct.Extensions)
+	if err != nil {
+		return nil, err
+	}
+	if len(extensionBytes) > 65535 {
+		return nil, fmt.Errorf("Extensions is too long: %d bytes", len(extensionBytes))
+	}
+	binary.Write(b, binary.BigEndian, uint16(len(extensionBytes)))
+	b.Write(extensionBytes)
+
+	// Signature
+	signatureBytes, err := base64.StdEncoding.DecodeString(sct.Signature)
+	if err != nil {
+		return nil, err
+	}
+	b.Write(signatureBytes)
+
+	return b.Bytes(), nil
+}
+
+var httpClient = http.Client{Timeout: 30 * time.Second}
+
+// Makes an HTTP request to the log server and returns the parsed SCT response.
+// If the cert is already in the log, the log will simply return the previously
+// generated SCT, making this idempotent.
+func submitSCT(url string, payload []byte) (*signedCertificateTimestamp, error) {
+	// Protocol must be HTTPS, so require people to omit it.
+	url = "https://" + url
+	if !strings.HasSuffix(url, "/") {
+		url = url + "/"
+	}
+
+	response, err := httpClient.Post(url+"ct/v1/add-chain", "application/json", bytes.NewReader(payload))
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP error %d", response.StatusCode)
+	}
+	sct := &signedCertificateTimestamp{}
+	err = json.NewDecoder(response.Body).Decode(&sct)
```

Comment:
This is the paranoia in me, but it should be easy to wrap `response.Body` with [`io.LimitReader`](https://golang.org/pkg/io/#LimitReader) using a sane size in case a CT server is compromised or buggy. I'm not sure what a reasonable size would be; it should definitely be large enough to handle any valid response we could possibly see, maybe times 2?